core/impl/src/main/java/org/jboss/jca/core/connectionmanager/pool/mcp/SemaphoreConcurrentLinkedDequeManagedConnectionPool.java

```diff
@@ -267,82 +282,92 @@ public void prefill()
-      }
-
-      subject = (subject == null) ? defaultSubject : subject;
-      cri = (cri == null) ? defaultCri : cri;
+         else if (debug)
+         {
+            String method = "getConnection(" + subject + ", " + cri + ")";
+            log.debug(ManagedConnectionPoolUtility.details(method, pool.getName(),
+                pool.getInternalStatistics().getInUseCount(), maxSize));
+         }
 
-      if (pool.isFull()) 
-      {
-         if (pool.getInternalStatistics().isEnabled())
-            pool.getInternalStatistics().deltaWaitCount();
+         subject = (subject == null) ? defaultSubject : subject;
+         cri = (cri == null) ? defaultCri : cri;
 
-         if (pool.isSharable() && (supportsLazyAssociation == null || supportsLazyAssociation.booleanValue())) 
+         if (pool.isFull())
          {
-            if (supportsLazyAssociation == null)
-               checkLazyAssociation();
+            if (pool.getInternalStatistics().isEnabled())
+               pool.getInternalStatistics().deltaWaitCount();
 
-            if (supportsLazyAssociation != null && supportsLazyAssociation.booleanValue()) 
+            if (pool.isSharable() && (supportsLazyAssociation == null || supportsLazyAssociation.booleanValue()))
             {
-               if (log.isTraceEnabled())
-                  log.tracef("Trying to detach - Pool: %s MCP: %s", pool.getName(), 
-                             Integer.toHexString(System.identityHashCode(this)));
+               if (supportsLazyAssociation == null)
+                  checkLazyAssociation();
 
-               if (!detachConnectionListener()) 
+               if (supportsLazyAssociation != null && supportsLazyAssociation.booleanValue())
                {
                   if (log.isTraceEnabled())
-                     log.tracef("Detaching didn't succeed - Pool: %s MCP: %s", pool.getName(), 
-                                Integer.toHexString(System.identityHashCode(this)));
+                     log.tracef("Trying to detach - Pool: %s MCP: %s", pool.getName(),
+                         Integer.toHexString(System.identityHashCode(this)));
+
+                  if (!detachConnectionListener())
+                  {
+                     if (log.isTraceEnabled())
+                        log.tracef("Detaching didn't succeed - Pool: %s MCP: %s", pool.getName(),
+                            Integer.toHexString(System.identityHashCode(this)));
+                  }
                }
             }
          }
-      }
 
-      long startWait = pool.getInternalStatistics().isEnabled() ? System.currentTimeMillis() : 0L;
-      try 
-      {
-         if (pool.getLock().tryAcquire(poolConfiguration.getBlockingTimeout(), TimeUnit.MILLISECONDS)) 
+         startWait = pool.getInternalStatistics().isEnabled() ? System.currentTimeMillis() : 0L;
+         if (pool.getLock().tryAcquire(poolConfiguration.getBlockingTimeout(), TimeUnit.MILLISECONDS))
```

Comment:
It is unclear to me where this lock meant to be released (other than when !isRunning()).

That said, this code also wrap its exclusive logic with try/finally blocks.
e.g.
```java
if (lock.tryAcquire(...)) {
   try {
       // exclusive logic
   } finally {
       lock.unlock();
   }
}

Reply:
This is a Semaphore used to grant configured number of connections from the pool, and not the thread synchronization object. It will be released either when error occurrs or durint returnConnection method. I agree that the naming is confusion and I have created JBJCA-1508 to change that.

Reply:
Ah - that explains it.